File: package/umt_python/tests/unit/object/test_remove_prototype_map_deep.py
import unittest

from src.object import remove_prototype_map_deep


class TestRemovePrototypeMapDeep(unittest.TestCase):
    def test_removes_polluting_properties_recursively_from_each(self):
        objects = [
            {"safe": {"__proto__": {"polluted": True}, "value": 1}},
            {
                "list": [{"prototype": {"polluted": True}, "ok": 1}],
                "meta": {"constructor": {"prototype": {"polluted": True}}, "keep": 2},
            },
        ]
        result = remove_prototype_map_deep(objects)
        self.assertEqual(
            result,
            [
                {"safe": {"value": 1}},
                {"list": [{"ok": 1}], "meta": {"keep": 2}},
            ],
        )
        self.assertNotIn("__proto__", result[0]["safe"])
        self.assertNotIn("prototype", result[1]["list"][0])
        self.assertNotIn("constructor", result[1]["meta"])

    def test_returns_empty_array_for_empty_input(self):
        self.assertEqual(remove_prototype_map_deep([]), [])

    def test_returns_new_list_instance(self):
        objects = [{"a": 1}]
        result = remove_prototype_map_deep(objects)
        self.assertIsNot(result, objects)
        self.assertIsInstance(result, list)

    def test_preserves_element_order(self):
        objects = [
            {"id": 1, "value": {"x": 1}},
            {"id": 2, "value": {"x": 2}},
            {"id": 3, "value": {"x": 3}},
        ]
        result = remove_prototype_map_deep(objects)
        self.assertEqual([item["id"] for item in result], [1, 2, 3])

    def test_does_not_mutate_input_objects(self):
        objects = [{"outer": {"__proto__": {"polluted": True}, "ok": 1}}]
        remove_prototype_map_deep(objects)
        self.assertIn("__proto__", objects[0]["outer"])
        self.assertEqual(objects[0]["outer"]["ok"], 1)

    def test_sanitizes_dangerous_keys_nested_inside_arrays(self):
        objects = [
            {
                "items": [
                    {"__proto__": {"polluted": True}, "a": 1},
                    {"constructor": {"polluted": True}, "b": 2},
                ]
            }
        ]
        result = remove_prototype_map_deep(objects)
        self.assertEqual(result[0]["items"], [{"a": 1}, {"b": 2}])

    def test_mixes_sanitized_and_already_safe_objects(self):
        objects = [
            {"clean": {"a": 1}},
            {"dirty": {"__proto__": {"polluted": True}, "b": 2}},
        ]
        result = remove_prototype_map_deep(objects)
        self.assertEqual(result, [{"clean": {"a": 1}}, {"dirty": {"b": 2}}])


if __name__ == "__main__":
    unittest.main()

File: package/umt_rust/tests/object/test_remove_prototype_map_deep.rs
use std::collections::HashMap;
use umt_rust::obj;
use umt_rust::object::{Value, umt_remove_prototype_map_deep};

fn as_map(value: Value) -> HashMap<String, Value> {
    value.into_object().unwrap()
}

#[test]
fn test_should_remove_prototype_polluting_properties_recursively_from_each_object() {
    let objects = vec![
        as_map(obj! {"safe" => obj!{"__proto__" => obj!{"polluted" => true}, "value" => 1}}),
        as_map(obj! {
            "list" => Value::Array(vec![obj!{"prototype" => obj!{"polluted" => true}, "ok" => 1}]),
            "meta" => obj!{"constructor" => obj!{"prototype" => obj!{"polluted" => true}}, "keep" => 2}
        }),
    ];

    let result = umt_remove_prototype_map_deep(&objects);

    let safe = result[0].get("safe").and_then(Value::as_object).unwrap();
    assert_eq!(safe.get("value"), Some(&Value::Int(1)));
    assert_eq!(safe.get("__proto__"), None);

    if let Some(Value::Array(list)) = result[1].get("list") {
        let item = list[0].as_object().unwrap();
        assert_eq!(item.get("ok"), Some(&Value::Int(1)));
        assert_eq!(item.get("prototype"), None);
    } else {
        panic!("list should be an array");
    }
    let meta = result[1].get("meta").and_then(Value::as_object).unwrap();
    assert_eq!(meta.get("keep"), Some(&Value::Int(2)));
    assert_eq!(meta.get("constructor"), None);
}

#[test]
fn test_should_return_an_empty_vector_when_given_an_empty_vector() {
    let objects: Vec<HashMap<String, Value>> = vec![];
    let result = umt_remove_prototype_map_deep(&objects);
    assert!(result.is_empty());
}

#[test]
fn test_should_preserve_element_order() {
    let objects = vec![
        as_map(obj! {"id" => 1, "value" => obj!{"x" => 1}}),
        as_map(obj! {"id" => 2, "value" => obj!{"x" => 2}}),
        as_map(obj! {"id" => 3, "value" => obj!{"x" => 3}}),
    ];

    let result = umt_remove_prototype_map_deep(&objects);

    let order: Vec<&Value> = result.iter().map(|item| item.get("id").unwrap()).collect();
    assert_eq!(order, vec![&Value::Int(1), &Value::Int(2), &Value::Int(3)]);
}

#[test]
fn test_should_sanitize_dangerous_keys_nested_inside_arrays() {
    let objects = vec![as_map(obj! {
        "items" => Value::Array(vec![
            obj!{"__proto__" => obj!{"polluted" => true}, "a" => 1},
            obj!{"constructor" => obj!{"polluted" => true}, "b" => 2},
        ])
    })];

    let result = umt_remove_prototype_map_deep(&objects);

    if let Some(Value::Array(items)) = result[0].get("items") {
        let item0 = items[0].as_object().unwrap();
        let item1 = items[1].as_object().unwrap();
        assert_eq!(item0.get("a"), Some(&Value::Int(1)));
        assert_eq!(item0.get("__proto__"), None);
        assert_eq!(item1.get("b"), Some(&Value::Int(2)));
        assert_eq!(item1.get("constructor"), None);
    } else {
        panic!("items should be an array");
    }
}

#[test]
fn test_should_handle_arrays_that_mix_sanitized_and_already_safe_objects() {
    let objects = vec![
        as_map(obj! {"clean" => obj!{"a" => 1}}),
        as_map(obj! {"dirty" => obj!{"__proto__" => obj!{"polluted" => true}, "b" => 2}}),
    ];

    let result = umt_remove_prototype_map_deep(&objects);

    assert_eq!(result[0].get("clean"), Some(&obj! {"a" => 1}));
    let dirty = result[1].get("dirty").and_then(Value::as_object).unwrap();
    assert_eq!(dirty.get("b"), Some(&Value::Int(2)));
    assert_eq!(dirty.get("__proto__"), None);
}
